File: AlgorithmsPackCPP/329. Longest Increasing Path in a Matrix.cpp
class Solution {
    int dr[4]{-1,0,1,0};
    int dc[4]{0,1,0,-1};
    
    struct node{
        int r;
        int c;
    };
    
    bool isValid(int r, int c, vector<vector<int>>&matrix){
        if(r < 0 || r>=matrix.size())
            return false;
        if(c<0||c>=matrix[0].size())
            return false;
        return true;
    }
public:
    pair<vector<node>,int>biggerNeighboursAndEdges(int r, int c, vector<vector<int>>&matrix){
        vector<node>edges;
        int incomingEdges = 0;
        
        for(int d = 0; d < 4; d++){
            int nr = dr[d]+r;
            int nc = dc[d]+c;
            if(!isValid(nr,nc,matrix))
                continue;
            
            if(matrix[r][c] < matrix[nr][nc])   //If the neighbouring node is bigger, we will push it into the 'biggerNeighbours' vector of nodes (used in the bfs below!)
                edges.push_back({nr,nc});  //3 5
            else if(matrix[r][c] > matrix[nr][nc])  //If the neighbouring node is smaller, it means there's an INCOMING EDGE to the original node!
                incomingEdges++;
        }
        return {edges,incomingEdges};   // This is return a vector of bigger neighbours, and the number of incoming edges to our node
    }
    
    int longestIncreasingPath(vector<vector<int>>& matrix) {
        int rows = matrix.size();
        int cols = matrix[0].size();
        vector<vector<int>>incomingEdges(rows,vector<int>(cols));
        queue<node>q;
        for(int r = 0; r < rows; r++){
            for(int c = 0; c < cols; c++){
                incomingEdges[r][c] = biggerNeighboursAndEdges(r,c,matrix).second;
                //So, all the convoluted calculations in the function above give us the number of incoming edges here!
                if(incomingEdges[r][c]==0)
                    q.push({r,c});  //We're adding all our potential 'starts' to the path into the queue!
                //Our bfs function will cleave off the starts...
            }
        }
        int levels = 0;
        while(!q.empty()){
            int level_sz = q.size();
            while(level_sz--){
                int r = q.front().r;
                int c = q.front().c;
                q.pop();
                vector<node>biggerNeighbours = biggerNeighboursAndEdges(r,c,matrix).first;
                
                for(auto&node:biggerNeighbours){
                    if(--incomingEdges[node.r][node.c]==0)
                        q.push(node);
                }
            }
            levels++;
        }
        return levels;
    }
};

File: AlgorithmsPackPythonSolutions/329. Longest Increasing Path in a Matrix.py
class Solution:
    dr = [-1, 0, 1, 0]
    dc = [0, 1, 0, -1]

    def isValid(self, r, c, matrix):
        if r < 0 or r >= len(matrix):
            return False
        if c < 0 or c >= len(matrix[0]):
            return False
        return True

    def biggerNeighboursAndEdges(self, r, c, matrix):
        edges = []
        incomingEdges = 0

        for d in range(4):
            nr = self.dr[d] + r
            nc = self.dc[d] + c
            if not self.isValid(nr, nc, matrix):
                continue

            if matrix[r][c] < matrix[nr][nc]:
                edges.append((nr, nc))
            elif matrix[r][c] > matrix[nr][nc]:
                incomingEdges += 1

        return edges, incomingEdges

    def longestIncreasingPath(self, matrix: List[List[int]]) -> int:
        rows = len(matrix)
        cols = len(matrix[0])
        incomingEdges = [[0] * cols for _ in range(rows)]
        q = []
        for r in range(rows):
            for c in range(cols):
                incomingEdges[r][c] = self.biggerNeighboursAndEdges(r, c, matrix)[1]
                if incomingEdges[r][c] == 0:
                    q.append((r, c))

        levels = 0
        while q:
            level_sz = len(q)
            while level_sz > 0:
                r, c = q.pop(0)
                level_sz -= 1
                biggerNeighbours = self.biggerNeighboursAndEdges(r, c, matrix)[0]

                for node in biggerNeighbours:
                    incomingEdges[node[0]][node[1]] -= 1
                    if incomingEdges[node[0]][node[1]] == 0:
                        q.append(node)

            levels += 1

        return levels
